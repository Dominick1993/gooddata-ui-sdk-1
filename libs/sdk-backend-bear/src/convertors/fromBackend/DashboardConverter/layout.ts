// (C) 2019-2021 GoodData Corporation

import { GdcDashboardLayout, GdcVisualizationObject, GdcVisualizationClass } from "@gooddata/api-model-bear";
import { uriRef, idRef, UriRef, areObjRefsEqual } from "@gooddata/sdk-model";
import {
    IFluidLayoutSize,
    IFluidLayoutSizeByScreen,
    IDashboardLayout,
    IDashboardLayoutColumn,
    IDashboardLayoutRow,
    IWidget,
    ResponsiveScreenType,
} from "@gooddata/sdk-backend-spi";
import { BearDashboardDependency } from "./types";

// Default layout column size for the kpi widget, when generating implicit layout
const KPI_SIZE = 2;
// Default layout column size for the visualization widget, when generating implicit layout
const VISUALIZATION_SIZE = 12;

export const convertResponsiveSize = (size: GdcDashboardLayout.IFluidLayoutSize): IFluidLayoutSize => {
    const converted: IFluidLayoutSize = {
        widthAsGridColumnsCount: size.width,
    };
    if (size.heightAsRatio) {
        converted.heightAsRatio = size.heightAsRatio;
    }

    return converted;
};

export const convertLayoutColumnSize = (
    column: GdcDashboardLayout.IFluidLayoutColSize,
): IFluidLayoutSizeByScreen => {
    const allScreens: ResponsiveScreenType[] = ["xl", "md", "lg", "sm", "xs"];

    return allScreens.reduce((acc: IFluidLayoutSizeByScreen, el) => {
        const size = column[el];
        if (size) {
            return {
                ...acc,
                [el]: convertResponsiveSize(size),
            };
        }

        return acc;
    }, {} as IFluidLayoutSizeByScreen);
};

const convertLayoutColumn = (
    column: GdcDashboardLayout.IFluidLayoutColumn,
    widgetDependencies: IWidget[],
): IDashboardLayoutColumn => {
    const { content } = column;
    if (GdcDashboardLayout.isLayoutWidget(content)) {
        const widget = widgetDependencies.find((dep) => {
            const { qualifier } = content.widget;
            if (GdcVisualizationObject.isObjUriQualifier(qualifier)) {
                return areObjRefsEqual(uriRef(qualifier.uri), dep.ref);
            }

            return areObjRefsEqual(idRef(qualifier.identifier), dep.ref);
        }) as IWidget;

        return {
            ...column,
            size: convertLayoutColumnSize(column.size),
            content: widget,
        };
    }

    return {
        content: content && convertLayout(content, widgetDependencies),
        size: convertLayoutColumnSize(column.size),
    };
};

const convertLayoutRow = (
    row: GdcDashboardLayout.IFluidLayoutRow,
    widgetDependencies: IWidget[],
): IDashboardLayoutRow => {
    return {
        ...row,
        columns: row.columns.map((column) => convertLayoutColumn(column, widgetDependencies)),
    };
};

export const convertLayout = (
    layout: GdcDashboardLayout.Layout,
    widgetDependencies: IWidget[],
): IDashboardLayout => {
    const {
        fluidLayout: { rows },
        fluidLayout,
    } = layout;
    const convertedLayout: IDashboardLayout = {
        type: "fluidLayout",
        rows: rows.map((row) => convertLayoutRow(row, widgetDependencies)),
    };
    if (fluidLayout.size) {
        convertedLayout.size = convertResponsiveSize(fluidLayout.size);
    }
    return convertedLayout;
};

/**
 * Create {@link ILegacyDashboardLayout} from {@link IWidget} items. As widgets do not contain any layout information,
 * implicit layout with a single row will be generated.
 *
 * @returns fluid layout created from the widgets
 */
export function createImplicitDashboardLayout(
    widgets: IWidget[],
    dependencies: BearDashboardDependency[],
    visualizationClasses: GdcVisualizationClass.IVisualizationClassWrapped[],
): IDashboardLayout | undefined {
    if (widgets.length < 1) {
        return undefined;
    }
    const rows = createRows(widgets, dependencies, visualizationClasses);

    return {
        type: "fluidLayout",
        rows,
    };
}

function createRows(
    widgets: IWidget[],
    dependencies: BearDashboardDependency[],
    visualizationClasses: GdcVisualizationClass.IVisualizationClassWrapped[],
): IDashboardLayoutRow[] {
    return [{ columns: createColumns(widgets, dependencies, visualizationClasses) }];
}

function createColumns(
    widgets: IWidget[],
    dependencies: BearDashboardDependency[],
    visualizationClasses: GdcVisualizationClass.IVisualizationClassWrapped[],
): IDashboardLayoutColumn[] {
    return widgets.map((widget) => createColumn(widget, dependencies, visualizationClasses));
}

function createColumn(
    widget: IWidget,
    dependencies: BearDashboardDependency[],
    visualizationClasses: GdcVisualizationClass.IVisualizationClassWrapped[],
): IDashboardLayoutColumn {
    return {
        content: widget,
        size: {
            xl: {
                widthAsGridColumnsCount: findImplicitWidgetWidth(widget, dependencies, visualizationClasses),
            },
        },
    };
}

function findImplicitWidgetWidth(
    widget: IWidget,
    dependencies: BearDashboardDependency[],
    visualizationClasses: GdcVisualizationClass.IVisualizationClassWrapped[],
) {
    if (widget.type === "kpi") {
        return KPI_SIZE;
    }

    const visualizationUri = (widget.insight as UriRef).uri;
    const vis = dependencies.find(
        (v) =>
            GdcVisualizationObject.isVisualization(v) && v.visualizationObject.meta.uri === visualizationUri,
    ) as GdcVisualizationObject.IVisualization;
    const visualizationClassUri = vis.visualizationObject.content.visualizationClass.uri;
    const visualizationClass = visualizationClasses.find(
        (visClass) => visClass.visualizationClass.meta.uri === visualizationClassUri,
    );
    return getVisualizationLegacyWidth(visualizationClass!);
}

function getVisualizationLegacyWidth(visualizationClass: GdcVisualizationClass.IVisualizationClassWrapped) {
    const visualizationType = visualizationClass.visualizationClass.content.url.split(":")[1];
    return visualizationType === "headline" ? KPI_SIZE : VISUALIZATION_SIZE;
}
