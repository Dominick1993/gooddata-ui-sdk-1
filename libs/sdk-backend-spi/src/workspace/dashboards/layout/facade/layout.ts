// (C) 2019-2021 GoodData Corporation
import invariant from "ts-invariant";
import { IFluidLayout, IFluidLayoutSize, isFluidLayout } from "../fluidLayout";

import { IFluidLayoutFacade, IFluidLayoutRowsMethods } from "./interfaces";
import { FluidLayoutRowsMethods } from "./rows";

/**
 * TODO: RAIL-2869 add docs
 * @alpha
 */
export class FluidLayoutFacade<TContent> implements IFluidLayoutFacade<TContent> {
    private constructor(private readonly layout: IFluidLayout<TContent>) {}

    /**
     * Creates an instance of FluidLayoutFacade
     * @param layout - layout to wrap
     */
    public static for<TContent>(layout: IFluidLayout<TContent>): IFluidLayoutFacade<TContent> {
        invariant(isFluidLayout(layout), "Provided data must be IFluidLayout!");
        return new FluidLayoutFacade(layout);
    }

    public rows = (): IFluidLayoutRowsMethods<TContent> => {
        return FluidLayoutRowsMethods.for(this, this.layout.rows);
    };

    public size = (): IFluidLayoutSize | undefined => {
        return this.layout.size;
    };

    public raw = (): IFluidLayout<TContent> => {
        return this.layout;
    };
}
