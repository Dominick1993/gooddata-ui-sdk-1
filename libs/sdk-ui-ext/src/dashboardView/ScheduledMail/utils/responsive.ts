// (C) 2019-2020 GoodData Corporation
import { SCREEN_BREAKPOINT_SM } from "../../../internal/dashboardEmbedding";

// https://github.com/gooddata/gdc-goodstrap/blob/master/src/core/MediaQueries.js#L61
export const isMobileView = (): boolean => window.innerWidth <= SCREEN_BREAKPOINT_SM;
