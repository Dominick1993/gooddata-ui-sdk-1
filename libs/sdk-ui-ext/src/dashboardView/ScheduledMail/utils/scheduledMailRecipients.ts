// (C) 2019-2020 GoodData Corporation
import { IScheduleEmailRecipient, isScheduleEmailExistingRecipient } from "../interfaces";

export const getScheduledEmailRecipientUniqueIdentifier = (recipient: IScheduleEmailRecipient): string =>
    isScheduleEmailExistingRecipient(recipient) ? recipient.user.login : recipient.email;

export const getScheduledEmailRecipientEmail = (recipient: IScheduleEmailRecipient): string =>
    isScheduleEmailExistingRecipient(recipient) ? recipient.user.email : recipient.email;

export const getScheduledEmailRecipientDisplayName = (recipient: IScheduleEmailRecipient): string =>
    isScheduleEmailExistingRecipient(recipient)
        ? `${recipient.user.firstName} ${recipient.user.lastName}`
        : recipient.email;
