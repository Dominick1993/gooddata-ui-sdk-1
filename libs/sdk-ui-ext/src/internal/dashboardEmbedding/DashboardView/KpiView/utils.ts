// (C) 2020-2021 GoodData Corporation
import {
    IAnalyticalBackend,
    IFilterContext,
    ITempFilterContext,
    IKpiWidget,
} from "@gooddata/sdk-backend-spi";
import {
    IFilter,
    IMeasure,
    IPoPMeasureDefinition,
    IPreviousPeriodMeasureDefinition,
    isAllTimeDateFilter,
    isDateFilter,
    isUriRef,
    newMeasure,
    newPopMeasure,
    newPreviousPeriodMeasure,
    ObjRef,
} from "@gooddata/sdk-model";
import {
    GoodDataSdkError,
    OnError,
    useBackend,
    useCancelablePromise,
    UseCancelablePromiseState,
    useWorkspace,
} from "@gooddata/sdk-ui";
import invariant from "ts-invariant";
import { filterContextToFiltersForWidget } from "../../converters";

interface IUseKpiDataConfig {
    kpiWidget: IKpiWidget;
    filterContext?: IFilterContext | ITempFilterContext;
    filters?: IFilter[];
    backend?: IAnalyticalBackend;
    workspace?: string;
    onError?: OnError;
}

interface IUseKpiDataResult {
    primaryMeasure: IMeasure;
    secondaryMeasure?: IMeasure<IPoPMeasureDefinition> | IMeasure<IPreviousPeriodMeasureDefinition>;
    filters: IFilter[];
}

export function useKpiData({
    kpiWidget,
    filters,
    filterContext,
    backend,
    workspace,
}: IUseKpiDataConfig): UseCancelablePromiseState<IUseKpiDataResult, GoodDataSdkError> {
    const effectiveBackend = useBackend(backend);
    const effectiveWorkspace = useWorkspace(workspace);

    const promise = async () => {
        invariant(kpiWidget.kpi, "The provided widget is not a KPI widget.");

        const filtersFromFilterContext = filterContextToFiltersForWidget(filterContext, kpiWidget);

        const inputFilters = [...filtersFromFilterContext, ...(filters ?? [])];

        const relevantFilters = await effectiveBackend
            .workspace(effectiveWorkspace)
            .dashboards()
            .getResolvedFiltersForWidget(kpiWidget, inputFilters);

        const primaryMeasure = newMeasure(kpiWidget.kpi.metric);

        const comparison = kpiWidget.kpi.comparisonType;

        const isAllTime =
            !relevantFilters ||
            !relevantFilters.some((filter) => isDateFilter(filter) && !isAllTimeDateFilter(filter));

        if (comparison === "none" || isAllTime) {
            return { primaryMeasure, filters: relevantFilters };
        }

        if (comparison === "previousPeriod") {
            const secondaryMeasure = newPreviousPeriodMeasure(primaryMeasure, [
                { dataSet: kpiWidget.dateDataSet, periodsAgo: 1 },
            ]);

            return { primaryMeasure, secondaryMeasure, filters: relevantFilters };
        }

        if (comparison === "lastYear") {
            const secondaryMeasure = await getLastYearComparisonMeasure(
                effectiveBackend,
                effectiveWorkspace,
                primaryMeasure,
                kpiWidget.dateDataSet,
            );

            return { primaryMeasure, secondaryMeasure, filters: relevantFilters };
        }

        invariant(false, `Unknown comparison ${comparison}`);
    };

    return useCancelablePromise({ promise }, [
        effectiveBackend,
        effectiveWorkspace,
        filters,
        filterContext,
        kpiWidget,
    ]);
}

async function getLastYearComparisonMeasure(
    backend: IAnalyticalBackend,
    workspace: string,
    primaryMeasure: IMeasure,
    targetDateDataset: ObjRef,
): Promise<IMeasure<IPoPMeasureDefinition>> {
    const catalog = await backend.workspace(workspace).catalog().forTypes(["dateDataset"]).load();
    const dateDatasets = catalog.dateDatasets();
    const relevantDateDataset = dateDatasets.find((dateDataset) => {
        if (isUriRef(targetDateDataset)) {
            return dateDataset.dataSet.uri === targetDateDataset.uri;
        } else {
            return dateDataset.dataSet.id === targetDateDataset.identifier;
        }
    });
    const yearAttribute = relevantDateDataset.dateAttributes.find(
        (dateAttribute) => dateAttribute.granularity === "GDC.time.year",
    );

    return newPopMeasure(primaryMeasure, yearAttribute.attribute.ref);
}
