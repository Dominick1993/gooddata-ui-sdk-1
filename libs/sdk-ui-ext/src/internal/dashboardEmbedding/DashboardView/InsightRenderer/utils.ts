// (C) 2020-2021 GoodData Corporation
import { IWidgetDefinition } from "@gooddata/sdk-backend-spi";
import { IFilter, newAllTimeFilter } from "@gooddata/sdk-model";
import { hasDateFilterForDateDataset } from "../../utils/filters";

export function addImplicitAllTimeFilter(widget: IWidgetDefinition, resolvedFilters: IFilter[]): IFilter[] {
    // if the widget is connected to a dateDataset and has no date filters for it in the context,
    // add an implicit All time filter for that dimension - this will cause the InsightView to ignore
    // any date filters on that dimension - this is how KPI dashboards behave
    if (widget.dateDataSet && !hasDateFilterForDateDataset(resolvedFilters, widget.dateDataSet)) {
        return [...resolvedFilters, newAllTimeFilter(widget.dateDataSet)];
    }
    return resolvedFilters;
}
