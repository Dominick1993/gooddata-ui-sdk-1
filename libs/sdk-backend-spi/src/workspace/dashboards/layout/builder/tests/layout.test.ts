// (C) 2019-2021 GoodData Corporation
import { FluidLayoutBuilder } from "../layout";
import { IFluidLayout, IFluidLayoutSize } from "../../fluidLayout";
import { createEmptyFluidLayoutBuilder, createValueOrUpdateCallbackTestCases } from "./utils";

describe("FluidLayoutBuilder", () => {
    describe("constructors", () => {
        describe(".for()", () => {
            it("should use the provided layout", () => {
                const layout: IFluidLayout<any> = {
                    type: "fluidLayout",
                    rows: [{ columns: [] }],
                };

                expect(FluidLayoutBuilder.for(layout).build()).toBe(layout);
            });

            it("should throw error, when the layout is damaged", () => {
                const boom = () => {
                    const layoud: any = {
                        type: "yolo",
                        rowz: [{ columnz: [] }],
                    };

                    FluidLayoutBuilder.for(layoud);
                };

                expect(boom).toThrowErrorMatchingSnapshot();
            });
        });

        describe(".forNewLayout()", () => {
            it("should create new empty layout", () => {
                const layout = FluidLayoutBuilder.forNewLayout().build();
                expect(layout).toMatchSnapshot();
            });
        });
    });

    describe(".size()", () => {
        const value: IFluidLayoutSize = { widthAsGridColumnsCount: 12, heightAsRatio: 50 };
        const testCases = createValueOrUpdateCallbackTestCases(value);
        it.each(testCases)("should update size %s", (_, valueOrUpdateCallback) => {
            const layout = createEmptyFluidLayoutBuilder().size(valueOrUpdateCallback).build();
            expect(layout).toMatchSnapshot();
        });
    });

    describe(".addRow()", () => {
        it("should add a row to the end of the layout by default", () => {
            const layoutBuilder = createEmptyFluidLayoutBuilder().addRow();

            const layoutBefore = layoutBuilder.build();
            const layoutAfter = layoutBuilder.addRow((r) => r.header({ title: "Added row" })).build();

            expect(layoutBefore).toMatchSnapshot("before");
            expect(layoutAfter).toMatchSnapshot("after");
        });

        it("should add a row to the provided index", () => {
            const layoutBuilder = createEmptyFluidLayoutBuilder().addRow().addRow();

            const layoutBefore = layoutBuilder.build();
            const layoutAfter = layoutBuilder.addRow((r) => r.header({ title: "Added row" }), 1).build();

            expect(layoutBefore).toMatchSnapshot("before");
            expect(layoutAfter).toMatchSnapshot("after");
        });
    });

    describe(".removeRow()", () => {
        it("should remove the row from the layout", () => {
            const rowIndex = 0;
            const layoutBuilder = createEmptyFluidLayoutBuilder().addRow(
                (r) => r.header({ title: "Row to remove" }),
                rowIndex,
            );

            const layoutBefore = layoutBuilder.build();
            const layoutAfter = layoutBuilder.removeRow(rowIndex).build();

            expect(layoutBefore).toMatchSnapshot("before");
            expect(layoutAfter).toMatchSnapshot("after");
        });

        it("should throw error, when the row at provided index does not exist", () => {
            const boom = () => {
                createEmptyFluidLayoutBuilder().removeRow(0);
            };

            expect(boom).toThrowErrorMatchingSnapshot();
        });
    });

    describe(".modifyRow()", () => {
        it("should modify the row", () => {
            const layoutBuilder = createEmptyFluidLayoutBuilder().addRow((r) =>
                r.header({ title: "Original row" }),
            );

            const layoutBefore = layoutBuilder.build();
            const layoutAfter = layoutBuilder
                .modifyRow(0, (r) => r.header({ title: "Modified row" }))
                .build();

            expect(layoutBefore).toMatchSnapshot("before");
            expect(layoutAfter).toMatchSnapshot("after");
        });

        it("should throw error, when the row at provided index does not exist", () => {
            const boom = () => {
                createEmptyFluidLayoutBuilder().modifyRow(0, (r) => r.header({ title: "Should throw" }));
            };

            expect(boom).toThrowErrorMatchingSnapshot();
        });
    });

    describe(".moveRow()", () => {
        it("should move the row", () => {
            const layoutBuilder = createEmptyFluidLayoutBuilder()
                .addRow((r) => r.header({ title: "Row 0" }))
                .addRow((r) => r.header({ title: "Row 1" }))
                .addRow((r) => r.header({ title: "Row 2" }));

            const layoutBefore = layoutBuilder.build();
            const layoutAfter = layoutBuilder.moveRow(0, 1).build();

            expect(layoutBefore).toMatchSnapshot("before");
            expect(layoutAfter).toMatchSnapshot("after");
        });

        it("should throw error, when the row at provided index does not exist", () => {
            const boom = () => {
                createEmptyFluidLayoutBuilder().moveRow(0, 1);
            };
            expect(boom).toThrowErrorMatchingSnapshot();
        });
    });

    describe(".removeRows()", () => {
        it("should remove all rows by default", () => {
            const layoutBuilder = createEmptyFluidLayoutBuilder().addRow().addRow();

            const layoutBefore = layoutBuilder.build();
            const layoutAfter = layoutBuilder.removeRows().build();

            expect(layoutBefore).toMatchSnapshot("before");
            expect(layoutAfter).toMatchSnapshot("after");
        });

        it("should remove selected rows only", () => {
            const layoutBuilder = createEmptyFluidLayoutBuilder()
                .addRow((r) => r.header({ title: "Row to keep 1" }))
                .addRow()
                .addRow((r) => r.header({ title: "Row to keep 2" }))
                .addRow();

            const layoutBefore = layoutBuilder.build();
            const layoutAfter = layoutBuilder
                .removeRows((rows) => rows.filter((r) => !r.hasHeader()))
                .build();

            expect(layoutBefore).toMatchSnapshot("before");
            expect(layoutAfter).toMatchSnapshot("after");
        });

        it("should remove selected row only", () => {
            const titleOfTheRowToRemove = "Row to remove";
            const layoutBuilder = createEmptyFluidLayoutBuilder()
                .addRow()
                .addRow((r) => r.header({ title: titleOfTheRowToRemove }))
                .addRow();

            const layoutBefore = layoutBuilder.build();
            const layoutAfter = layoutBuilder
                .removeRows((rows) => rows.find((r) => r.titleEquals(titleOfTheRowToRemove)))
                .build();

            expect(layoutBefore).toMatchSnapshot("before");
            expect(layoutAfter).toMatchSnapshot("after");
        });

        it("should not remove any row, when selection result is undefined", () => {
            const layoutBuilder = createEmptyFluidLayoutBuilder().addRow();

            const layoutBefore = layoutBuilder.build();
            const layoutAfter = layoutBuilder
                .removeRows((rows) => rows.find((r) => r.titleEquals("Non-existing row title")))
                .build();

            expect(layoutBefore).toMatchSnapshot("before");
            expect(layoutAfter).toMatchSnapshot("after");
        });
    });

    describe(".modifyRows()", () => {
        it("should modify all rows by default", () => {
            const layoutBuilder = createEmptyFluidLayoutBuilder().addRow().addRow();

            const layoutBefore = layoutBuilder.build();
            const layoutAfter = layoutBuilder
                .modifyRows((r, f) => r.header({ title: `Modified row ${f.index()}` }))
                .build();

            expect(layoutBefore).toMatchSnapshot("before");
            expect(layoutAfter).toMatchSnapshot("after");
        });

        it("should modify selected rows only", () => {
            const layoutBuilder = createEmptyFluidLayoutBuilder()
                .addRow((r) => r.header({ title: "Original row" }))
                .addRow()
                .addRow((r) => r.header({ title: "Original row" }))
                .addRow();

            const layoutBefore = layoutBuilder.build();
            const layoutAfter = layoutBuilder
                .modifyRows(
                    (r, f) => r.header({ title: `Modified row ${f.index()}` }),
                    (rows) => rows.filter((r) => r.hasHeader()),
                )
                .build();

            expect(layoutBefore).toMatchSnapshot("before");
            expect(layoutAfter).toMatchSnapshot("after");
        });

        it("should modify selected row only", () => {
            const titleOfTheRowToModify = "Original row";
            const layoutBuilder = createEmptyFluidLayoutBuilder()
                .addRow()
                .addRow((r) => r.header({ title: titleOfTheRowToModify }))
                .addRow();

            const layoutBefore = layoutBuilder.build();
            const layoutAfter = layoutBuilder
                .modifyRows(
                    (r) => r.header({ title: "Modified row" }),
                    (rows) => rows.find((r) => r.titleEquals(titleOfTheRowToModify)),
                )
                .build();

            expect(layoutBefore).toMatchSnapshot("before");
            expect(layoutAfter).toMatchSnapshot("after");
        });

        it("should not modify any row, when selection result is undefined", () => {
            const layoutBuilder = createEmptyFluidLayoutBuilder().addRow();

            const layoutBefore = layoutBuilder.build();
            const layoutAfter = layoutBuilder
                .modifyRows(
                    (r) => r.header({ title: "That would never happen" }),
                    (rows) => rows.find((r) => r.titleEquals("Non-existing row title")),
                )
                .build();

            expect(layoutBefore).toMatchSnapshot("before");
            expect(layoutAfter).toMatchSnapshot("after");
        });
    });
});
