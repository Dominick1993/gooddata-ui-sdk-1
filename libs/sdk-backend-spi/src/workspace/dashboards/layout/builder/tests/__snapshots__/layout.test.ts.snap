// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FluidLayoutBuilder .addRow() should add a row to the end of the layout by default: after 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
    },
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Added row",
      },
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .addRow() should add a row to the end of the layout by default: before 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .addRow() should add a row to the provided index: after 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
    },
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Added row",
      },
    },
    Object {
      "columns": Array [],
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .addRow() should add a row to the provided index: before 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
    },
    Object {
      "columns": Array [],
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .modifyRow() should modify the row: after 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Modified row",
      },
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .modifyRow() should modify the row: before 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Original row",
      },
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .modifyRow() should throw error, when the row at provided index does not exist 1`] = `"Cannot modify the row - row at index 0 does not exist!"`;

exports[`FluidLayoutBuilder .modifyRows() should modify all rows by default: after 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Modified row 0",
      },
    },
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Modified row 1",
      },
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .modifyRows() should modify all rows by default: before 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
    },
    Object {
      "columns": Array [],
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .modifyRows() should modify selected row only: after 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
    },
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Modified row",
      },
    },
    Object {
      "columns": Array [],
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .modifyRows() should modify selected row only: before 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
    },
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Original row",
      },
    },
    Object {
      "columns": Array [],
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .modifyRows() should modify selected rows only: after 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Modified row 0",
      },
    },
    Object {
      "columns": Array [],
    },
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Modified row 2",
      },
    },
    Object {
      "columns": Array [],
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .modifyRows() should modify selected rows only: before 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Original row",
      },
    },
    Object {
      "columns": Array [],
    },
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Original row",
      },
    },
    Object {
      "columns": Array [],
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .modifyRows() should not modify any row, when selection result is undefined: after 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .modifyRows() should not modify any row, when selection result is undefined: before 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .moveRow() should move the row: after 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Row 1",
      },
    },
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Row 0",
      },
    },
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Row 2",
      },
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .moveRow() should move the row: before 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Row 0",
      },
    },
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Row 1",
      },
    },
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Row 2",
      },
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .moveRow() should throw error, when the row at provided index does not exist 1`] = `"Cannot move the row - row at index 0 does not exist!"`;

exports[`FluidLayoutBuilder .removeRow() should remove the row from the layout: after 1`] = `
Object {
  "rows": Array [],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .removeRow() should remove the row from the layout: before 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Row to remove",
      },
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .removeRow() should throw error, when the row at provided index does not exist 1`] = `"Cannot remove the row - row at index 0 does not exist!"`;

exports[`FluidLayoutBuilder .removeRows() should not remove any row, when selection result is undefined: after 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .removeRows() should not remove any row, when selection result is undefined: before 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .removeRows() should remove all rows by default: after 1`] = `
Object {
  "rows": Array [],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .removeRows() should remove all rows by default: before 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
    },
    Object {
      "columns": Array [],
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .removeRows() should remove selected row only: after 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
    },
    Object {
      "columns": Array [],
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .removeRows() should remove selected row only: before 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
    },
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Row to remove",
      },
    },
    Object {
      "columns": Array [],
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .removeRows() should remove selected rows only: after 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Row to keep 1",
      },
    },
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Row to keep 2",
      },
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .removeRows() should remove selected rows only: before 1`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Row to keep 1",
      },
    },
    Object {
      "columns": Array [],
    },
    Object {
      "columns": Array [],
      "header": Object {
        "title": "Row to keep 2",
      },
    },
    Object {
      "columns": Array [],
    },
  ],
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .size() should update size by callback 1`] = `
Object {
  "rows": Array [],
  "size": Object {
    "heightAsRatio": 50,
    "widthAsGridColumnsCount": 12,
  },
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .size() should update size by callback returning undefined 1`] = `
Object {
  "rows": Array [],
  "size": undefined,
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .size() should update size by undefined 1`] = `
Object {
  "rows": Array [],
  "size": undefined,
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder .size() should update size by value 1`] = `
Object {
  "rows": Array [],
  "size": Object {
    "heightAsRatio": 50,
    "widthAsGridColumnsCount": 12,
  },
  "type": "fluidLayout",
}
`;

exports[`FluidLayoutBuilder constructors .for() should throw error, when the layout is damaged 1`] = `"Provided data must be IFluidLayout!"`;

exports[`FluidLayoutBuilder constructors .forNewLayout() should create new empty layout 1`] = `
Object {
  "rows": Array [],
  "type": "fluidLayout",
}
`;
